catalog/cataloger_diff.go

```diff
@@ -175,17 +333,152 @@ func getDiffDifferences(ctx context.Context, tx db.Tx, limit int, after string)
+		}
+
+		diffType := evaluateFromChildElementDiffType(effectiveCommits, parentID, childEnt, parentEnt)
+		if diffType == DifferenceTypeNone {
+			continue
+		}
+
+		err = batch.Write(&diffResultRecord{
+			SourceBranch: childID,
+			DiffType:     diffType,
+			Entry:        childEnt.Entry,
+			EntryCtid:    &childEnt.RowCtid,
+		})
+		if err != nil {
+			return err
+		}
+
+		// stop on limit
+		records++
+		if limit > -1 && records >= limit {
+			break
+		}
+	}
+	if err := childScanner.Err(); err != nil {
+		return err
+	}
+	return batch.Flush()
+}
+
+func createDiffResultsTable(ctx context.Context, executor sq.Execer) (string, error) {
+	diffResultsTableName, err := diffResultsTableNameFromContext(ctx)
+	if err != nil {
+		return "", err
+	}
+	_, err = executor.Exec("CREATE UNLOGGED TABLE " + diffResultsTableName + ` (
+		source_branch bigint NOT NULL,
+		diff_type integer NOT NULL,
+		path character varying COLLATE "C" NOT NULL,
+		entry_ctid tid
+	)`)
+	if err != nil {
+		return "", fmt.Errorf("diff from child diff result table: %w", err)
+	}
+	return diffResultsTableName, nil
+}
 
+func evaluateFromChildElementDiffType(effectiveCommits *diffEffectiveCommits, parentBranchID int64, childEnt *DBScannerEntry, parentEnt *DBScannerEntry) DifferenceType {
+	var matchedParent *DBScannerEntry
+	if parentEnt != nil && parentEnt.Path == childEnt.Path {
+		matchedParent = parentEnt
+	}
+	// when the entry was deleted
+	if childEnt.IsDeleted() {
+		if matchedParent == nil || matchedParent.IsDeleted() {
+			return DifferenceTypeNone
+		}
+		if matchedParent.BranchID == parentBranchID {
+			// check if parent did any change to the entity
+			if matchedParent.MinCommit > effectiveCommits.ParentEffectiveCommit {
+				return DifferenceTypeConflict
+			}
+			return DifferenceTypeRemoved
+		}
+		// check if the parent saw this entry at the time
+		if matchedParent.MinCommit >= effectiveCommits.ParentEffectiveCommitByBranchID(matchedParent.BranchID) {
+			return DifferenceTypeRemoved
+		}
+		return DifferenceTypeNone
+	}
+	// when the entry was not deleted
+	if matchedParent != nil {
+		if matchedParent.IsDeleted() {
+			return DifferenceTypeAdded
+		}
+		if matchedParent.BranchID == parentBranchID {
+			// check if parent did any change to the entity
+			if matchedParent.MinCommit > effectiveCommits.ParentEffectiveCommit {
+				return DifferenceTypeConflict
+			}
+			return DifferenceTypeChanged
+		}
+		if matchedParent.MinCommit >= effectiveCommits.ParentEffectiveCommitByBranchID(matchedParent.BranchID) {
+			return DifferenceTypeChanged
+		}
+		return DifferenceTypeAdded
+	}
+
+	return DifferenceTypeAdded
+}
+
+func insertDiffResultsBatch(exerciser sq.Execer, tableName string, batch []*diffResultRecord) error {
+	if len(batch) == 0 {
+		return nil
+	}
+	ins := psql.Insert(tableName).Columns("source_branch", "diff_type", "path", "entry_ctid")
+	for _, rec := range batch {
+		ins = ins.Values(rec.SourceBranch, rec.DiffType, rec.Entry.Path, rec.EntryCtid)
```

Comment:
I don't understand.  You write a CTID of another table into a _table_?  Then everything up to the top-level exported functions has to carry a comment saying it is only useful inside a transaction (with an appropriately high isolation level), otherwise this is unsafe in the presence of concurrent VACUUMing FULL.  At the very least, add warning lines to all system documentation never to VACUUM FULL the tables.
Similarly, this imposes immediate restrictions either on any implementation of metadata retention (which deletes entries) or of all code that uses the result -- the CTID might point at nothing.
Also, please document the restrictions on restoring from backups: I assume most backup methods will have to invalidate ctid on restore, so the result of writing these diff entries must be trashed during backup/restore.

The manual [has this to say](https://www.postgresql.org/docs/12/ddl-system-columns.html):
> ctid
>
>    The physical location of the row version within its table. Note that although the ctid can be used to locate the row version very quickly, a row's ctid will change if it is updated or moved by VACUUM FULL. Therefore ctid is useless as a long-term row identifier. A primary key should be used to identify logical rows.

Our reliance on CTIDs is becoming a danger to data integrity.  This technical debt might be acceptable because currently there is no breakage that you or I can see.  But at the very least it requires precise exact documentation.

Reply:
This is the same mechanism as today - the cataloger's Merge calls the diff implementation inside the merge transaction and for entries that needs to update/added as part of the diff, we use the ctid to select the full entry information.